starlette/applications.py

```diff
@@ -139,8 +140,8 @@ def add_middleware(
 
     def add_exception_handler(
         self,
-        exc_class_or_status_code: int | typing.Type[Exception],
```

Comment:
NB: this has now been tightened so that the Exception for `exc_class_or_status_code` must match the Exception in the handler Callable.

Reply:
@Kludex isn't this a deprecated function?

Reply:
No. We never deprecated `add_exception_handler`... 🤔

Reply:
BTW, to be more precise, the Exception in the handler Callable is still contravariant, so this passes as one would hope:

```python
def fallback_exception_handler(request: Request, exc: Exception) -> JSONResponse:
    ...

app.add_exception_handler(Exception, fallback_exception_handler)
app.add_exception_handler(MyException, fallback_exception_handler)
```

ie: the `ExceptionType` TypeVar changes the handler Callable `Exception` to a type parameter. The value of this type parameter is provided at the call site, eg: in the example here its `Exception` in the first call, and `MyException` in the second. But the contravariance of Callable remains.

And also the Callable is still not covariant, so this fails as one might expect:

```python
def my_exception_handler(request: Request, exc: MyException) -> JSONResponse:
    ...

# this fails, because subtypes of the exc_class_or_status_code Exception in the Callable 
# aren't permitted ie: Callable is not covariant in the function's input parameter types
app.add_exception_handler(Exception, my_exception_handler)
```